numpy/_core/src/common/npy_hashtable.cpp

```diff
@@ -240,3 +240,20 @@ PyArrayIdentityHash_GetItem(PyArrayIdentityHash *tb, PyObject *const *key)
     PyObject *res = find_item(tb, key)[0];
     return res;
 }
+
+#ifdef Py_GIL_DISABLED
```

Comment:
I wonder whether a separate function is needed at all. I.e., could one instead just define `PyArrayIdentityHash_GetItem` differently for the with and without GIL case? (have an `#ifdef !Py_GIL_DISABLED` above, and an `#else` here). That way, the rest of the code does not have to worry about locking at all. (I presume CPython does something similarly, since I don't think there is a separate `PyDict_GetItem` for the no-GIL case).

Reply:
Sure, that would work too. Then callers would only have to care about the mutex if they need to write to the cache, which should only ever happen inside the dispatching implementation where it happens right now.

Reply:
> do the same for *SetItem, so that all the code elsewhere does not have to worry about the locking any more. Would that work?

The `SetItem` call is only ever used in the dispatching implementation. IMO I'd rather not do this refactor for `SetItem` unless there's a customer for mutating the dispatch cache outside the dispatching implementation. (There's one use in `_multiarray.tests.c.src`, but IMO that's not worth worrying about)

Reply:
Just using a regular dict wouldn't help either. The thread safety problem in this cache isn't that the data structure isn't atomic, it's that the whole recursive ufunc promotion algorithm isn't atomic. As soon as there's a cache miss there needs to be some kind of lock preventing multiple threads from computing the value that fills the cache simultaneously.

It's definitely a subtle point: this code went through several rounds of iteration. I even highlighted it in the PyCon talk I co-presented this year as a spot that caused a multithreaded scaling issue. The original implementation we shipped used a critical section  instead of the shared_mutex and that scaled poorly.